### ISSUE TITLE
Double-change bug with FileWatching on OS X

### ISSUE THREAD

**Issue by yurivish** [JuliaLang/julia#25880]
Watching a file with `watch_file` reports the file changed twice — once immediately upon changing, and another one around a second later. I think this might have something to do with the timing of the underlying IO system writing to the filesystem, since this only happens if the saved file is non-empty; if you save an empty file, then it generates a single immediate change notification and nothing further.

```
# watch.jl
using FileWatching

file = joinpath(@__DIR__, "file.txt")
counter = 0

while true
    event = watch_file(file)
    if event.changed
        global counter += 1
        @show counter
        println("file changed.")
    end
end
```

Put that in a file called `watch.jl` and create another file `file.txt` in the same directory. Then run `julia watch.jl`, wait for it to start up, then modify and save `file.txt`. Here's what I see in the console for a single change:

```
counter = 1
file changed.
counter = 2
file changed.
```

This was initially observed in the context of Revise (https://github.com/timholy/Revise.jl/issues/83#issuecomment-362940187), and @timholy has run into seemingly related mysteries in the past: https://github.com/timholy/Revise.jl/issues/15, https://github.com/timholy/Revise.jl/pull/22

**Comment by Keno** [comment]
This behavior is just what we get from the underlying file system APIs. E.g. the same happens in Node.js (e.g. https://stackoverflow.com/questions/8246933/detect-mtime-resolution-of-filesystem). I don't think there's anything actionable for us here. If the OS developers tighten the semantics, that'd be great, but unless and until then all we can do is report it. Our docs already link to the Node.js docs on this, which mentions this behavior.

**Comment by vtjnash** [comment]
There is a bug in libuv (I can't find the tracker item for it right now) where it closes and reopens all open uv_fs_event watchers every time you call `watch_file`, which can miss or duplicate events that were queued at the time. It is kinda hard to use watch_file correctly since it is vulnerable to a race at the start (watch_folder is better at this).

**Comment by Keno** [comment]
Ok, but that doesn't quite seem like this issue? Can we have a separate issue for that with the proper upstream reference?

**Comment by vtjnash** [comment]
Yeah, seems likely a bit different (also currently only hypothetical since we haven't had someone report it). I was unable to find the upstream reference right now.

**Comment by vtjnash** [comment]
https://github.com/libuv/libuv/issues/3866
